### PR TITLE
fix: Make sure that each customer can only create one review per product

### PR DESCRIPTION
**Pull request by aragon999**
### 1. Why is this change necessary?
Currently the validation is not correct, such that the `id` of the product review will be compared with the customer id. Therefore it is possible to create multiple reviews per product (if one is omitting the id, which usually does not happen in the normal storefront). Furthermore it would be possible to overwrite the name or email from the request, which also does not happen.

### 2. What does this change do, exactly?
Fix the validation and always use the values from the customer object and cleanup some other stuff.

### 3. Describe each step to reproduce the issue or behaviour.
\-

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by leduc92** [comment]
Hi @aragon999 , thanks for your contribution. I can understand the benefits for some customers. If we adopt this functionality, we would like to offer the merchant a choice in the form of a setting. I would suggest to add the setting on Settings >  Products

**Comment by aragon999** [comment]
@leduc92 a setting for what exactly do you mean? :-)

At the moment, the storefront does not offer the functionality, to create multiple reviews. Once the review is created, it is only possible to edit the review and the same customer cannot add another review. It is only possible to add a second review, when one is directly using the route, and only because of a bug in the route. So I am not sure, because adding multiple reviews, would also require some more frontend work.

**Comment by aragon999** [comment]
@leduc92 any update?

**Comment by leduc92** [comment]
Hi, @aragon999 , sorry for being late in my response. I mean there's a switcher setting in the administration says "Customer's able to review your product once" for example, it's disabled by default.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php [hunk 1]
```diff
@@ -52,37 +51,30 @@ public function getDecorated(): AbstractProductReviewSaveRoute
...
-        $languageId = $context->getContext()->getLanguageId();
-        $salesChannelId = $context->getSalesChannel()->getId();
+        // Should always be valid, as the route requires login
+        \assert($customer !== null);
 
-        $customerId = $customer->getId();
+        $data->set('customerId', $customer->getId());
+        $data->set('name', $customer->getFirstName());
+        $data->set('lastName', $customer->getLastName());
+        $data->set('email', $customer->getEmail());
```
Comment (ghost) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php]:
```suggestion
```
We can now remove this 3 here and directly set `'name' => $customer->getFirstName()` & `'email' => $customer->getEmail()` in the `$review` object. `lastName` is unused and can be removed

**Comment by ghost** [comment]
> Hi, @aragon999 , sorry for being late in my response. I mean there's a switcher setting in the administration says "Customer's able to review your product once" for example, it's disabled by default.

@leduc92 currently it is not intended to create multiple reviews. There is no switch in the administration for this.
See e.g. https://github.com/shopware/shopware/blob/trunk/src/Storefront/Page/Product/Review/ProductReviewLoader.php#L116

Multiple parts in the core & storefront are designed to only handle a single review per product per customer.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php (hunk 1 above)
Comment (aragon999) [reply src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php]: I will change that, and probably rebase the whole PR :-) Thank you

**Comment by aragon999** [comment]
@leduc92 yes, @akf-bw is right. It would require some major rework. And this PR is "just" fixing a bug. Although it might be useful to have multiple reviews (however I cannot think of a good usecase).

**Comment by aragon999** [comment]
@leduc92 ping.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php [hunk 2]
```diff
@@ -52,39 +51,28 @@ public function getDecorated(): AbstractProductReviewSaveRoute
...
-        $languageId = $context->getContext()->getLanguageId();
-        $salesChannelId = $context->getSalesChannel()->getId();
-
-        $customerId = $customer->getId();
-
-        if (!$data->has('name')) {
-            $data->set('name', $customer->getFirstName());
+        $salesChannelId = $context->getSalesChannelId();
+        if ($this->config->getBool('core.listing.showReview', $salesChannelId) === false) {
+            throw new ReviewNotActiveExeption();
```
Comment (phuoccx) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php:56]:
```suggestion
            throw ProductException::reviewNotActive();
```

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php [hunk 3]
```diff
@@ -52,39 +51,28 @@ public function getDecorated(): AbstractProductReviewSaveRoute
...
-        if (!$data->has('lastName')) {
-            $data->set('lastName', $customer->getLastName());
-        }
+        $customer = $context->getCustomer();
 
-        if (!$data->has('email')) {
-            $data->set('email', $customer->getEmail());
-        }
+        // Should always be valid, as the route requires login
+        \assert($customer !== null);
```
Comment (phuoccx) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php:62]:
```suggestion
        \assert($customer instanceof CustomerEntity);
```

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php [hunk 4]
```diff
@@ -52,39 +51,28 @@ public function getDecorated(): AbstractProductReviewSaveRoute
...
+            throw new ReviewNotActiveExeption();
         }
 
-        if (!$data->has('lastName')) {
-            $data->set('lastName', $customer->getLastName());
-        }
+        $customer = $context->getCustomer();
 
-        if (!$data->has('email')) {
-            $data->set('email', $customer->getEmail());
```
Comment (phuoccx) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php:74]: We dispatched ReviewFormEvent, which contains data such as name, last name, and email. However, you removed it from the data. I believe this change may cause a breaking change for those who subscribe to the ReviewFormEvent.

File: src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php [hunk 5]
```diff
@@ -120,54 +107,38 @@ private function validate(DataBag $data, Context $context): void
...
-            $criteria->addFilter(new EqualsFilter('customerId', $data->get('customerId')));
-            $criteria->addFilter(new EqualsFilter('productId', $data->get('productId')));
+            $reviewValidationCriteria->addFilter(new EqualsFilter('productId', $data->get('productId')));
 
             $definition->add('customerId', new EntityNotExists([
                 'entity' => 'product_review',
                 'context' => $context,
-                'criteria' => $criteria,
+                'criteria' => $reviewValidationCriteria,
+                'primaryProperty' => 'customerId',
```
Comment (phuoccx) [inline src/Core/Content/Product/SalesChannel/Review/ProductReviewSaveRoute.php:131]: We can not add \`primaryProperty\` here. That should follow the type of the $option at EntityNotExists.php::__construct()

So please update the type-hint as well

File: src/Core/Framework/DataAbstractionLayer/Validation/EntityExists.php [hunk 6]
```diff
@@ -35,10 +35,7 @@ class EntityExists extends Constraint
      */
     public function __construct(array $options)
     {
-        $options = array_merge(
-            ['criteria' => new Criteria()],
-            $options
-        );
+        $options['criteria'] ??= new Criteria();
```
Comment (phuoccx) [inline src/Core/Framework/DataAbstractionLayer/Validation/EntityExists.php:38]: Could you explain why this change is necessary? I agree your change is for better performance, but Is it not the root cause of the bug?

**Comment by aragon999** [comment]
@phuoccx mh okay, by now the core parts, which would be modified by this PR have been changed a lot since when I originally created the PR, so rebasing and cleaning it up will be quite annoying and a challenge... -.-

But I will look into it.

By now the feature is more broken then before my PR: https://github.com/shopware/shopware/commit/9d3f958af71dd0b0ffb5b79f4444acb976a6a0a2#diff-42314a9ee71e000524f2a02479ef6dc3f44afe8255fa63c7f818ffac9bdaa29e

**Comment by leduc92** [comment]
Thank you for your contribution to our project! I noticed that this pull request has been inactive for a while, so I'm going to close it for now to keep our repo tidy. Please feel free to reopen it whenever you're ready to continue working on it or need further review. We appreciate your involvement and look forward to your future contributions!
